@pepopowitz => [Autosuggest] Start letting keyboard navigate preview pane

mzikherman opened the pull request:
cc @damassi 

Marked this as WIP but it might be mergeable (w/ possibly some cleanup?). One thing we need is how to render the 'selected' state of preview items. For now I just went with a beautiful green background.

![input](https://user-images.githubusercontent.com/1457859/53444581-919bb580-39dc-11e9-86b2-c30511b644a6.gif)

It's probably worth pulling this down and playing with it, but basically you get left/right behavior working as you might expect, and then hitting 'enter' will navigate you to a preview item (if one is selected). Otherwise, hitting 'enter' will navigate you to the selected suggestion item.

Essentially the approach is:
  - autosuggest container listens for arrow keypresses and sets that in state
  - preview item lists can read that state and know which preview item is 'selected' and can highlight it
  - preview item lists register the list of possible hrefs, so autosuggest container can navigate to the correct one

![input](https://user-images.githubusercontent.com/1457859/53444752-fc4cf100-39dc-11e9-93e7-723aa3aacdaa.gif)

damassi commented on src/Components/Search/Previews/Grids/ArtistSearch/MarketingCollections.tsx:
```diff
@@ -54,50 +57,82 @@ export const CollectionTitles = ({ title }: { title: string }) => {
-      )
-    }
-  )
+    this.props.searchState.addHrefs(hrefs)
```
Might be a little clearer to rename `addHrefs` to `registerHrefs`

damassi commented on src/Components/Search/SearchBar.tsx:
```diff
@@ -315,25 +320,67 @@ export class SearchBar extends Component<Props, State> {
+          const { state } = searchState
+          return (
+            <AutosuggestContainer
+              onKeyDown={e => {
```
Would probably be a little more tidy to move this into a `handleKeyboardNavigation` method; or better -- rename  `AutosuggestContainer` to `AutosuggestManager` and move it into its own file, which hides away this keyboard handler logic as well as state manipulation. Then can do 

```jsx
<AutosuggestManager>
  <Autosuggest ... />
</AutosuggestManager>
```
and keep the tree clean.

mzikherman replied:
Interesting! I like that. Will experiment with some better structure here.

pepopowitz replied:
+1 it'd be nice to extract this somewhere.

damassi commented on src/Components/Search/SearchBar.tsx:
```diff
@@ -315,25 +320,67 @@ export class SearchBar extends Component<Props, State> {
+                  return this.renderSuggestionsContainer(props, { xs })
+                }}
+                inputProps={inputProps}
+                onSuggestionSelected={(e, selection) => {
```
`onSuggestionSelected={this.handleSuggestionSelected}`

damassi commented on src/Components/Search/state.ts:
```diff
@@ -0,0 +1,48 @@
+    this.setState({ hasEnteredPreviews, selectedPreviewIndex })
+  }
+
+  addHrefs(hrefs) {
```
Rather than thinking in terms of `hrefs` maybe think in terms of items? 

```
registerPreviewItem(item)
```
and then later, navigate via `item.href`.

damassi commented on src/Components/Search/state.ts:
```diff
@@ -0,0 +1,48 @@
+    this.setState({ previewHrefs: hrefs })
+  }
+
+  clearHrefs() {
```
similarly, `clearPreviewItems`

damassi commented on src/Components/Search/SearchBar.tsx:
```diff
@@ -315,25 +320,67 @@ export class SearchBar extends Component<Props, State> {
+                  } else {
+                    searchState.enterPreview()
+                  }
+                } else if (e.keyCode === 37 && state.hasEnteredPreviews) {
```
```js
const keys = {
  DOWN: 39,
  UP: 40 
}

... 

if (e.keyCode === keys.UP) { ... } 
```

damassi commented on src/Components/Search/SearchBar.tsx:
```diff
@@ -315,25 +320,67 @@ export class SearchBar extends Component<Props, State> {
+
+                  if (
+                    state.hasEnteredPreviews &&
+                    state.previewHrefs[state.selectedPreviewIndex]
```
```
const selectedIndex = state.previewHrefs[state.selectedPreviewIndex]
```

pepopowitz commented on src/Components/Search/state.ts:
```diff
@@ -0,0 +1,48 @@
+
+  incrementSelectedPreviewIndex() {
+    this.setState({
+      selectedPreviewIndex: this.state.selectedPreviewIndex + 1,
```
Do we need to guard against overflowing the index?

mzikherman replied:
good call, we should. Will add the total count into this state so we can accomplish that.

mzikherman replied:
Oh nevermind all the items for a preview pane are registered so can just do `items.length`!

damassi commented on src/Components/Search/AutosuggestManager.tsx:
```diff
@@ -0,0 +1,82 @@
+      }}
+    </Subscribe>
+  )
+})
```
Looks great

damassi commented:
> FWIW - if we replace `unstated` with `useContext`, I don't think it will look a whole lot different than what you've got here.

Agree. Might be worth merging and then immediately following up, just to get unstated out sooner rather than later.

mzikherman commented:
Ok @pepopowitz I think this is ready for re-review/merge. I addressed the comments, most notable are extracting the `AutosuggestManager`, and also registering item objects vs just the hrefs. Also removed the green highlighting and it matches the comps.

I think this is mergeable.

Re: unstated vs hooks- I'd love for that to be a follow up item!

artsyit commented:
<!-- GITHUB_RELEASE COMMENT: released -->
:rocket: PR was released in v14.1.15 :rocket:
